mlflow/store/s3_artifact_repo.py

```diff
@@ -69,15 +69,29 @@ def list_artifacts(self, path=None):
         for result in results:
             # Subdirectories will be listed as "common prefixes" due to the way we made the request
             for obj in result.get("CommonPrefixes", []):
-                subdir = obj.get("Prefix")[len(artifact_path)+1:]
-                if subdir.endswith("/"):
-                    subdir = subdir[:-1]
-                infos.append(FileInfo(subdir, True, None))
+                subdir_path = obj.get("Prefix")
+                if not subdir_path.startswith(artifact_path):
+                    raise ValueError(
+                        "The path of the listed S3 directory does not begin with the specified"
+                        " artifact path. Artifact path: {artifact_path}. Subdirectory path:"
+                        " {subdir_path}.".format(
+                            artifact_path=artifact_path, subdir_path=subdir_path))
+                subdir_rel_path = self.get_path_module().relpath(
+                    path=subdir_path, start=artifact_path)
+                if subdir_rel_path.endswith("/"):
+                    subdir_rel_path = subdir_rel_path[:-1]
+                infos.append(FileInfo(subdir_rel_path, True, None))
             # Objects listed directly will be files
             for obj in result.get('Contents', []):
-                name = obj.get("Key")[len(artifact_path)+1:]
-                size = int(obj.get('Size'))
-                infos.append(FileInfo(name, False, size))
+                file_path = obj.get("Key")
+                if not file_path.startswith(artifact_path):
+                    raise ValueError(
+                        "The path of the listed S3 file does not begin with the specified"
+                        " artifact path. Artifact path: {artifact_path}. File path:"
```

Comment:
Is there any way this duplicated code can be unified? I know this is extracted differently - one is a file vs directory. In fact, if there is a way share some of this between different blob stores -- would be awesome .... so any changes to internal methods don't require changes all over the place.

Reply:
For now, I've deduped the code inside `S3ArtifactRepository` by defining a static `_verify_listed_object_contains_artifact_path_prefix()` method. It may be reasonable to create an abstract class for blob stores that follow this pattern, but it seems we agree that this may be a larger undertaking that requires a followup PR.